### PR TITLE
boards: move programmer includes in Makefile.include

### PR DESCRIPTION
**Pull request by aabadie**
### Contribution description

This PR is a refactoring of the way board programmer variables are set. The current logic is to include the corresponding makefile (openocd.inc.mk, jlink.inc.mk) for each board and to eventually manage some logic there depending on the programmer used.

The idea of this PR is to include the programmer logic once in Makefile.include and to only specify the related variables in the boards (`PROGRAMMER`, `OPENOCD_CONFIG`, `DEBUG_ADAPTER`, etc).

This saves a few redundant code and avoid the inclusion of programmer variable if they are not needed (if not using flash/debug targets for instance).

Some changes were applied to the avrdude management where there is also a PROGRAMMER used but with a different meaning: it's related to the avrdude sub-programmer and not the main programmer (avrdude itself).

If this PR makes sense, we could also apply the same logic with `serial.inc.mk`

### Testing procedure

- Test that all boards can still be flashed and debugged (tried on arduino-mkr1000, samr21-xpro, nucleo-l412kb but we need more :) )
- Green Murdock

### Issues/PRs references

Had this idea while reviewing #15459 (in https://github.com/RIOT-OS/RIOT/pull/15459#discussion_r525507044 exactly)

### CODE REVIEW

**Comment by aabadie** [comment]
I hope I didn't forget something, so review with care :)

File: Makefile.include [hunk 1]
```diff
@@ -395,6 +395,11 @@ include $(BOARDDIR)/Makefile.include
 INCLUDES += -I$(RIOTCPU)/$(CPU)/include
 include $(RIOTCPU)/$(CPU)/Makefile.include
 
+# Include programmer logic if available and if flash/debug/debug-server is used
+ifneq (,$(filter flash% debug% riotboot/flash%,$(MAKECMDGOALS)))
```
Comment (aabadie) [inline Makefile.include:399]:
```suggestion
ifneq (,$(filter reset flash% debug% riotboot/flash%,$(MAKECMDGOALS)))
```

Comment (fjmolinas) [inline Makefile.include:399]:
```suggestion
ifneq (,$(filter flash% debug% riotboot/flash% riotboot/bootloader/flash%,$(MAKECMDGOALS)))
```

There is also

File: makefiles/boards/stm32.inc.mk [hunk 2]
```diff
@@ -17,9 +17,6 @@ else
   PORT_DARWIN ?= $(firstword $(sort $(wildcard /dev/tty.usbmodem*)))
 endif
 
-# setup serial terminal
-include $(RIOTMAKE)/tools/serial.inc.mk
-
```
Comment (fjmolinas) [inline makefiles/boards/stm32.inc.mk:22]: Unrelated?

**Review by fjmolinas — COMMENTED** [review]
Looks good, some small comments above. Should the dfu renaming be in its own commit? I'll run some tests.

**Comment by aabadie** [comment]
Thanks for the comments, will address them shortly. I thought that it would also be good to check that the use if DEBUG_ADAPTER_ID is still working.

**Comment by fjmolinas** [comment]
Some commands are now yielding the following error:

```
BOARD=slstk3402a make -C tests/bloom_bytes/ info-modules --no-print-directory
/home/francisco/workspace/RIOT/tests/bloom_bytes/../../Makefile.include:548: *** FLASHFILE is not defined for this board: .  Stop.
```

```
 *** FLASHFILE is not defined for this board: .  Stop.
```

**Comment by aabadie** [comment]
> Some commands are now yielding the following error

Good catch. Do you have any suggestion of how this should be fixed ?

**Comment by aabadie** [comment]
> Do you have any suggestion of how this should be fixed ?

I could add `info-module` to the list of targets allowed to in include the board programmer code. What do you think ?

**Comment by fjmolinas** [comment]
> FLASHFILE is not defined for this board:

And `FLASHFILE` is used by the `fuzz` target, and is used by native as well, `test-murdock`

I'm also seeing that some `BOARD`s do not use `PROGRAMMER` but directly set the `FLASHER`, could that be fixed so they all use `PROGRAMMER`?

**Comment by fjmolinas** [comment]
> > Do you have any suggestion of how this should be fixed ?
> 
> I could add `info-module` to the list of targets allowed to include the board programmer code. What do you think ?

Its not only info module, its `info-%` everything. IMO setting the flashfile should be done in function of the target as well, don't think it is needed for the `info-target`.

**Comment by aabadie** [comment]
Another possibility is to include the programmer include unconditionally, this is the same as now, but only done in one place. That would solve the FLASHFILE issue.

> I'm also seeing that some BOARDs do not use PROGRAMMER but directly set the FLASHER, could that be fixed so they all use PROGRAMMER?

The include is only done if the file exists so that _should_ harmless. Otherwise, variables related to these programmers could be moved to new include files (I noticed that `lpc2k_pgm` is used in several places)

File: makefiles/boards/sam0.inc.mk [hunk 3]
```diff
@@ -43,14 +43,8 @@ ifeq ($(PROGRAMMER),)
 endif
 
 ifeq ($(PROGRAMMER),edbg)
-  # use edbg for flashing
+  # use edbg for flashing. must be included before openocd
```
Comment (aabadie) [inline makefiles/boards/sam0.inc.mk:46]: Here I think that `openocd.inc.mk` could be included at the end of `edbg.inc.mk` instead of here. And then we could drop all the logic used here.
That would work because: FLASHER will defined first for edbg and DEBUGGER/DEBUGSERVER variables will be defined after by code in openocd.inc.mk. FLASHER won't be reset since it uses `?=` in `openocd.inc.mk`.

Comment (aabadie) [reply makefiles/boards/sam0.inc.mk]: Did that after testing on samr21-xpro, already pushed in a separate commit.

**Comment by fjmolinas** [comment]
> Another possibility is to include the programmer include unconditionally, this is the same as now, but only done in one place. That would solve the FLASHFILE issue.
> 
> > I'm also seeing that some BOARDs do not use PROGRAMMER but directly set the FLASHER, could that be fixed so they all use PROGRAMMER?
> 
> The include is only done if the file exists so that _should_ harmless. Otherwise, variables related to these programmers could be moved to new include files (I noticed that `lpc2k_pgm` is used in several places)

I'm fine with this a s a first step.

**Comment by fjmolinas** [comment]
Flashed the following BOARDs with no issues:

```
arduino-mega2560 arduino-uno arduino-zero b-l072z-lrwan1 ek-lm4f120xl frdm-k64f frdm-kw41z i-nucleo-lrwan1 nrf52840dk nrf52dk nucleo-f091rc nucleo-f103rb nucleo-f303k8 nucleo-f303re nucleo-f334r8 nucleo-f767zi nucleo-l073rz nucleo-l152re nucleo-l433rc nucleo-l496zg p-nucleo-wb55 pba-d-01-kw2x remote-revb samr21-xpro slstk3402a z1
```

**Comment by fjmolinas** [comment]
Please squash @aabadie

**Comment by aabadie** [comment]
squashed!

**Review by fjmolinas — APPROVED** [review]
Retested all seems good. ACK.
